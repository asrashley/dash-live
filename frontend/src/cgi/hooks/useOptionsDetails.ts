import { type ReadonlySignal, useSignal } from "@preact/signals";

import { fieldGroups } from "@dashlive/options";

import { InputOptionName } from "../types/InputOptionName";

function createOptionNames(): InputOptionName[] {
    const names: InputOptionName[] = [];
    fieldGroups.forEach(grp => {
        grp.fields.forEach(({ name, fullName, shortName }) => {
            names.push({
                cgiName: name,
                fullName,
                shortName,
            });
        });
    });
    return names;
}

export interface UseOptionsDetailsHook {
    allOptions: ReadonlySignal<InputOptionName[]>;
}

export function useOptionsDetails(): UseOptionsDetailsHook {
    const allOptions = useSignal<InputOptionName[]>(createOptionNames());
    return { allOptions };
}