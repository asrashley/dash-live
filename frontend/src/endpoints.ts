import { createContext } from 'preact';
import log from 'loglevel';

import { routeMap, uiRouteMap} from '@dashlive/routemap';
import { CsrfTokenStore } from './CsrfTokenStore';
import { CsrfTokenCollection } from './types/CsrfTokenCollection';
import { AllManifests } from './types/AllManifests';
import { AllStreamsJson, AllStreamsResponse } from './types/AllStreams';
import { JWToken } from './types/JWToken';
import { MultiPeriodStream, MultiPeriodStreamJson } from './types/MultiPeriodStream';
import { DecoratedMultiPeriodStream } from "./types/DecoratedMultiPeriodStream";
import { MultiPeriodStreamSummary } from './types/MultiPeriodStreamSummary';
import { ContentRolesMap } from './types/ContentRolesMap';
import { ModifyMultiPeriodStreamJson, ModifyMultiPeriodStreamResponse } from './types/ModifyMultiPeriodStreamResponse';
import { InitialApiTokens } from './types/InitialApiTokens';
import { LoginRequest } from './types/LoginRequest';
import { LoginResponse } from './types/LoginResponse';
import { MultiPeriodStreamValidationRequest, MultiPeriodStreamValidationResponse } from './types/MpsValidation';
import { CgiOptionDescription } from './types/CgiOptionDescription';

type TokenStoreCollection = {
  files: CsrfTokenStore;
  kids: CsrfTokenStore;
  login: CsrfTokenStore;
  streams: CsrfTokenStore;
  upload: CsrfTokenStore;
};

export type ApiRequestOptions = {
  authorization?: string;
  body: RequestInit['body'];
  service: keyof TokenStoreCollection;
  signal?: AbortSignal;
  method: RequestInit['method'];
  rejectOnError?: boolean;
  query?: URLSearchParams;
};

export type GetAllStreamsProps = ApiRequestOptions & {
  withDetails: boolean
};

type RefreshAccessTokenResponse = {
  accessToken?: JWToken;
  csrfTokens: CsrfTokenCollection;
  ok: boolean;
  status: number;
};

export interface ApiRequestsProps extends Readonly<InitialApiTokens> {
  navigate: (url: string) => void;
}

export class ApiRequests {
  private csrfTokens: TokenStoreCollection;
  private accessToken: JWToken | null;
  private refreshToken: JWToken | null;
  private navigate: ApiRequestsProps['navigate'];

  constructor({csrfTokens, accessToken, refreshToken, navigate}: ApiRequestsProps) {
    this.csrfTokens = {
      files: new CsrfTokenStore(csrfTokens?.files),
      kids: new CsrfTokenStore(csrfTokens?.kids),
      login: new CsrfTokenStore(csrfTokens?.login),
      streams: new CsrfTokenStore(csrfTokens?.streams),
      upload: new CsrfTokenStore(csrfTokens?.upload),
    }
    this.accessToken = accessToken;
    this.refreshToken = refreshToken;
    this.navigate = navigate;
  }

  getAllManifests(options: Partial<ApiRequestOptions> = {}): Promise<AllManifests> {
    return this.sendApiRequest<AllManifests>(routeMap.listManifests.url(), options);
  }

  getContentRoles(options: Partial<ApiRequestOptions> = {}): Promise<ContentRolesMap> {
    return this.sendApiRequest(routeMap.contentRoles.url(), options);
  }

  getCgiOptions(options: Partial<ApiRequestOptions> = {}) : Promise<CgiOptionDescription[]> {
    return this.sendApiRequest(routeMap.cgiOptions.url(), options);
  }

  async getUserInfo(signal: AbortSignal): Promise<LoginResponse | Response> {
    if (!this.refreshToken) {
      throw new Error('Cannot get user info without a refresh token');
    }
    const options: Partial<ApiRequestOptions> = {
      authorization: this.refreshToken.jwt,
      method: 'GET',
      rejectOnError: false,
      signal,
    };
    const response = await this.sendApiRequest<LoginResponse | Response>(routeMap.login.url(), options);
    if (response['success']) {
      this.accessToken = (response as LoginResponse).accessToken;
    }
    return response;
  }

  async loginUser(request: LoginRequest, options: Partial<GetAllStreamsProps> = {}): Promise<LoginResponse> {
    const response: LoginResponse = await this.sendApiRequest<LoginResponse>(routeMap.login.url(), {
      body: JSON.stringify(request),
      method: 'POST',
      service: 'login',
      ...options,
    });
    if (response.success) {
      if (response.accessToken) {
        this.accessToken = response.accessToken;
      }
      if (response.refreshToken) {
        this.refreshToken = response.refreshToken;
      }
    }
    return response;
  }

  async logoutUser(options: Partial<GetAllStreamsProps> = {}): Promise<Response> {
    const response: Response = await this.sendApiRequest(routeMap.login.url(), {
      method: 'DELETE',
      service: 'login',
      ...options,
    });
    this.accessToken = null;
    this.refreshToken = null;
    return response;
  }

  async getAllStreams(options: Partial<GetAllStreamsProps> = {}): Promise<AllStreamsResponse> {
    const { streams, keys } = await this.sendApiRequest<AllStreamsJson>(routeMap.listStreams.url(), options);
    return { streams, keys};
  }

  getAllMultiPeriodStreams(options: Partial<ApiRequestOptions> = {}): Promise<MultiPeriodStreamSummary[]> {
    return this.sendApiRequest<MultiPeriodStreamSummary[]>(routeMap.listMps.url(), options);
  }

  async getMultiPeriodStream(mps_name: string, options: Partial<ApiRequestOptions> = {}): Promise<MultiPeriodStream> {
    const { model } = await this.sendApiRequest<MultiPeriodStreamJson>(routeMap.editMps.url({mps_name}), options);
    return model;
  }

  async addMultiPeriodStream(
      data: DecoratedMultiPeriodStream,
      options: Partial<ApiRequestOptions> = {}): Promise<ModifyMultiPeriodStreamResponse> {
    const service = 'streams';
    const csrf_token = await this.csrfTokens[service].getToken(
      options?.signal, this.getCsrfTokens);
    const { errors, success, model } = await this.sendApiRequest<ModifyMultiPeriodStreamJson>(routeMap.addMps.url(), {
      ...options,
      service,
      body: JSON.stringify({
        ...data,
        csrf_token
      }),
      method: 'PUT',
    });
    return { errors, success, model };
  }

  async modifyMultiPeriodStream(mps_name: string, data: DecoratedMultiPeriodStream,
     options: Partial<ApiRequestOptions> = {}): Promise<ModifyMultiPeriodStreamJson> {
    const service = 'streams';
    const csrf_token = await this.csrfTokens[service].getToken(
      options?.signal, this.getCsrfTokens);
    return await this.sendApiRequest<ModifyMultiPeriodStreamJson>(routeMap.editMps.url({mps_name}), {
      ...options,
      service,
      body: JSON.stringify({...data, csrf_token}),
      method: 'POST',
    });
  }

  async deleteMultiPeriodStream(mps_name: string, options: Partial<ApiRequestOptions> = {}): Promise<Response> {
    const service = 'streams';
    const csrf_token = await this.csrfTokens[service].getToken(
      options?.signal, this.getCsrfTokens
    );
    const query = new URLSearchParams({
      csrf_token,
    });
    return this.sendApiRequest<Response>(routeMap.editMps.url({mps_name}), {
      ...options,
      service,
      query,
      method: 'DELETE',
    });
  }

  async validateMultiPeriodStream(data: MultiPeriodStreamValidationRequest,
        options: Partial<ApiRequestOptions> = {}): Promise<MultiPeriodStreamValidationResponse> {
    const service = 'streams';
    const csrf_token = await this.csrfTokens[service].getToken(
      options?.signal, this.getCsrfTokens);
    return this.sendApiRequest<MultiPeriodStreamValidationResponse>(routeMap.validateMps.url(), {
      ...options,
      method: 'POST',
      service: 'streams',
      body: JSON.stringify({...data, csrf_token}),
    });
  }

  private async sendApiRequest<T>(url: string, options: Partial<ApiRequestOptions>): Promise<T> {
    const { authorization, body, service, signal, method='GET',
      rejectOnError = true } = options;
    let { query } = options;
    let usedAccessToken = false;
    const headers: Headers = new Headers({
      "Content-Type": "application/json",
    });
    if (authorization !== undefined) {
      headers.set('Authorization', `Bearer ${authorization}`);
    } else {
      if (!this.accessToken && this.refreshToken) {
        await this.getAccessToken(signal);
      }
      if (this.accessToken) {
        headers.set('Authorization', `Bearer ${this.accessToken.jwt}`);
        usedAccessToken = true;
      } else if (service) {
        const token = await this.csrfTokens[service].getToken(signal);
        if (query === undefined) {
          query = new URLSearchParams({csrf_token: token});
        } else {
          query.set('csrf_token', token);
        }
      }
    }
    if (query) {
      url = `${url}?${query.toString()}`;
    }
    if (signal && signal.aborted) {
      throw signal.reason;
    }
    const cache = 'no-cache';
    const credentials = 'same-origin';
    const mode = 'same-origin';
    url = new URL(url, document.location.origin).href;
    log.trace(`sendApiRequest(${url})`);
    let fetchResult = await fetch(url, {
      cache,
      credentials,
      mode,
      body,
      headers,
      method,
      signal,
    });
    log.trace(`url=${url} status=${fetchResult.status} usedAccessToken=${usedAccessToken}`);
    if (fetchResult.status === 401 && usedAccessToken && this.refreshToken) {
      await this.getAccessToken(signal);
      headers.set('Authorization', `Bearer ${this.accessToken.jwt}`);
      log.trace(`retrying sendApiRequest(${url})`);
      fetchResult = await fetch(url, {
          cache,
          credentials,
          mode,
          body,
          headers,
          method,
          signal,
      });
    }
    if (!fetchResult.ok) {
      if (rejectOnError) {
        throw new Error(`${ url }: ${ fetchResult.status }`);
      } else {
        return fetchResult as T;
      }
    }
    if (fetchResult.status !== 200) {
      return fetchResult as T;
    }
    const data = await fetchResult.json();
    if (signal?.aborted) {
      throw signal.reason;
    }
    if (typeof(data?.csrfTokens) === "object") {
      this.updateCsrfTokens(data.csrfTokens);
    } else if(typeof(data?.csrf_tokens) === "object") {
      this.updateCsrfTokens(data.csrf_tokens);
    } else if (typeof(data?.csrf_token) === "string") {
      this.updateCsrfTokens({[service]: data.csrf_token});
    }

    return data as T;
  }

  private updateCsrfTokens(tokens: Partial<CsrfTokenCollection>) {
    for (const [key, value] of Object.entries(tokens)) {
      if (this.csrfTokens[key] === undefined) {
        this.csrfTokens[key] = new CsrfTokenStore(value);
      } else {
        this.csrfTokens[key].setToken(value);
      }
    }
  }

  private async getAccessToken(signal: AbortSignal): Promise<JWToken> {
    if (!this.refreshToken) {
      throw new Error('Cannot request an access token without a refresh token');
    }
    const options = {
      authorization: this.refreshToken.jwt,
      rejectOnError: false,
      signal,
    };
    const data = await this.sendApiRequest<RefreshAccessTokenResponse>(routeMap.refreshAccessToken.url(), options);
    const { accessToken, csrfTokens, ok, status } = data ?? {};
    if (ok === false && status === 401) {
      this.navigate(uiRouteMap.login.url());
    }
    if (!accessToken) {
      throw new Error('Failed to refresh access token');
    }
    this.accessToken = accessToken;
    if (csrfTokens) {
      this.updateCsrfTokens(csrfTokens);
    }
    return accessToken;
  }

  private getCsrfTokens = async (signal: AbortSignal): Promise<void> => {
    if (!this.accessToken && this.refreshToken) {
      await this.getAccessToken(signal);
    }
    if (!this.accessToken) {
      throw new Error('Cannot request CSRF tokens without an access token');
    }
    const options = {
      authorization: this.accessToken.jwt,
      signal,
    };
    const data = await this.sendApiRequest<RefreshAccessTokenResponse>(routeMap.refreshCsrfTokens.url(), options);
    const { csrfTokens } = data ?? {};
    if (csrfTokens) {
      this.updateCsrfTokens(csrfTokens);
    }
  };
}

export const EndpointContext = createContext<ApiRequests>(null);
