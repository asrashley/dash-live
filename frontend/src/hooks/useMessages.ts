import { type ReadonlySignal, signal, useComputed } from "@preact/signals";
import { MessageLevel, MessageType } from "../types/MessageType";

type MessagesState = {
  alerts: MessageType[];
  nextId: number;
}

const messages = signal<MessagesState>({ alerts: [], nextId: 1 });

// only exported for use in tests
export function resetAllMessages() {
  messages.value = { alerts: [], nextId: 1 };
}

export type AppendMessageFn = (level: MessageLevel, text: string) => number;

const appendMessage: AppendMessageFn = (level: MessageLevel, text: string) => {
  const id = messages.value.nextId;
  const alerts = [
    ...messages.value.alerts,
    {
      text,
      id,
      level,
    },
  ];
  messages.value = {
    nextId: id + 1,
    alerts,
  };
  return id;
};

const removeAlert = (id: number) => {
  const alerts = messages.value.alerts.filter((a) => a.id !== id);
  messages.value = { ...messages.value, alerts };
};

export interface UseMessagesHook {
  alerts: ReadonlySignal<MessageType[]>
  appendMessage: AppendMessageFn;
  removeAlert: (id: number) => void;
}

export function useMessages(): UseMessagesHook {
  const alerts = useComputed(() => {
    return messages.value.alerts;
  });

  return { alerts, appendMessage, removeAlert };
}
