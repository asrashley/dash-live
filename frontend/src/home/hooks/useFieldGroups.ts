import { useContext } from "preact/hooks";
import { type ReadonlySignal, useComputed } from "@preact/signals";

import { fieldGroups, drmSystems } from '@dashlive/options';

import { useAllManifests  } from '../../hooks/useAllManifests';
import { UseCombinedStreams  } from '../../hooks/useCombinedStreams';
import { StaticInputProps } from "../../form/types/StaticInputProps";
import { InputFormGroup } from "../../form/types/InputFormGroup";

const drmSkipNames = new RegExp(`^${drmSystems.map(name => `${name}__enabled`).join('|')}$`);

interface UseFieldGroupsHook {
  homeFieldGroups: ReadonlySignal<InputFormGroup[]>;
}

export function useFieldGroups(): UseFieldGroupsHook {
  const { allManifests, names } = useAllManifests();
  const { streamNames, streamsMap } = useContext(UseCombinedStreams);

  const homeFieldGroups = useComputed<InputFormGroup[]>(() => {
    const playbackMode: StaticInputProps = {
      name: "mode",
      shortName: "mode",
      fullName: "playbackMode",
      title: "Playback Mode",
      type: "radio",
      options: [
        {
          title: "Video On Demand (using live profile)",
          value: "vod",
        },
        {
          title: "Live stream (using live profile)",
          value: "live",
        },
        {
          title: "Video On Demand (using on-demand profile)",
          value: "odvod",
        },
      ],
    };
    const selectManifest: StaticInputProps = {
      name: "manifest",
      shortName: "manifest",
      fullName: "manifest",
      title: "Manifest",
      text: "Manifest template to use",
      type: "select",
      options: names.value.map((name: string) => {
        const msft = allManifests.value[name];
        return {
          title: msft.title,
          value: name,
        };
      }),
    };
    const selectStream: StaticInputProps = {
      name: "stream",
      shortName: "stream",
      fullName: "stream",
      title: "Stream",
      text: "Stream to play",
      type: "select",
      options: streamNames.value.map((value: string) => ({
        title: streamsMap.value.get(value).title,
        value,
      })),
    };
    const selectDrmSystem: StaticInputProps = {
      name: "drms",
      shortName: "drms",
      fullName: "drms",
      title: "DRM systems",
      text: "DRM systems to enable for encrypted AdaptationSets",
      type: "multiselect",
      options: drmSystems.map((name: string) => ({
        name,
        title: name,
        value: name,
      })),
    };
    const generalOptions: InputFormGroup = {
      ...fieldGroups[0],
      fields: [
        playbackMode,
        selectManifest,
        selectStream,
        selectDrmSystem,
        ...fieldGroups[0].fields,
      ],
    };
    const otherGroups: InputFormGroup[] = fieldGroups.slice(1).map(grp => ({
        ...grp,
        fields: grp.fields.filter(({name}) => !drmSkipNames.test(name)),
    }));
    return [generalOptions, ...otherGroups];
  });

  return { homeFieldGroups };
}
