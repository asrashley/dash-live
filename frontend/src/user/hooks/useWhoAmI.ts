import { createContext } from "preact";
import { useCallback } from "preact/hooks";
import { type ReadonlySignal, useComputed, useSignal } from "@preact/signals";

import { UserState } from "../types/UserState";
import { InitialUserState } from "../types/InitialUserState";
import { FlattenedUserState } from "../types/FlattenedUserState";
import { flattenUserState } from "../utils/flattenUserState";

export interface UseWhoAmIHook {
  user: ReadonlySignal<UserState>;
  flattened: ReadonlySignal<FlattenedUserState>;
  setUser: (ius: InitialUserState | null) => void;
}

export const WhoAmIContext = createContext<UseWhoAmIHook>(null);

const blankState: InitialUserState = {
  mustChange: false,
  lastLogin: null,
  groups:[],
};

export function useWhoAmI(): UseWhoAmIHook {
  const userInfo = useSignal<InitialUserState>(structuredClone(blankState));
  const user = useComputed<UserState>(() => ({
    ...userInfo.value,
    isAuthenticated: userInfo.value.pk !== undefined,
    permissions: {
      admin: userInfo.value.groups.includes('ADMIN'),
      media: userInfo.value.groups.includes('MEDIA'),
      user: userInfo.value.groups.includes('USER'),
    },
  }));
  const flattened = useComputed<FlattenedUserState>(() => flattenUserState(user.value));

  const setUser = useCallback((ius: InitialUserState | null) => {
    if (ius) {
      userInfo.value = structuredClone(ius);
    } else {
      userInfo.value = structuredClone(blankState);
    }
  }, [userInfo]);

  return { user, flattened, setUser };
}
