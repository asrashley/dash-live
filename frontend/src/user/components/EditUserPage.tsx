import { useCallback, useContext } from "preact/hooks";
import { useComputed, useSignal } from "@preact/signals";
import { useLocation, useParams } from "wouter-preact";

import { uiRouteMap } from "@dashlive/routemap";

import { FlattenedUserState, useAllUsers, UserValidationErrors } from "../../hooks/useAllUsers";
import { RouteParamsType } from "../../types/RouteParamsType";
import { EditUserCard } from "./EditUserCard";
import { AppStateContext } from "../../appState";
import { DeleteUserDialog } from "./DeleteUserDialog";
import { EndpointContext } from "../../endpoints";
import { useMessages } from "../../hooks/useMessages";

export default function EditUserPage() {
  const [, setLocation] = useLocation();
  const { dialog, closeDialog } = useContext(AppStateContext);
  const apiRequests = useContext(EndpointContext);
  const { appendMessage } = useMessages();
  const { flattenedUsers, error, validateUser, updateUser } = useAllUsers();
  const { username } = useParams<RouteParamsType>();
  const changes = useSignal<Partial<FlattenedUserState>>({});
  const userEntry = useComputed<FlattenedUserState | undefined>(() =>
    flattenedUsers.value.find((user) => user.username === username)
  );
  const user = useComputed(() => {
    const usr: FlattenedUserState = {
      adminGroup: false,
      mediaGroup: false,
      userGroup: false,
      mustChange: true,
      lastLogin: null,
      ...userEntry.value,
      ...changes.value,
    };
    return usr;
  });
  const errors = useComputed<UserValidationErrors>(() => validateUser(user.value));

  const setValue = useCallback(
    (field: string, value: string | number | boolean) => {
      changes.value = {
        ...changes.value,
        [field]: value,
      };
    },
    [changes]
  );

  const saveChanges = useCallback(async () => {
    try {
      const result = await apiRequests.editUser(user.value);
      if (result.success) {
        appendMessage("success", `User ${username} successfully modified`);
        updateUser(result.user);
        setLocation(uiRouteMap.listUsers.url());
      } else {
        appendMessage("warning", `Failed to modify user ${username}`);
      }
    } catch(err) {
      appendMessage("warning", `Failed to modify user ${username} - ${err}`);
    }
  }, [apiRequests, appendMessage, setLocation, updateUser, user, username]);

  const requestDelete = useCallback(() => {
    dialog.value = {
      backdrop: true,
      confirmDelete: {
        name: username,
        confirmed: false,
      },
    };
  }, [dialog, username]);

  const onConfirmDelete = useCallback(async () => {
    try {
      const result = await apiRequests.deleteUser(user.value.pk);
      if (result.ok) {
        appendMessage("success", `User ${username} successfully deleted`);
        setLocation(uiRouteMap.listUsers.url());
      } else {
        appendMessage("warning", "Failed to delete user");
      }
    } catch (err) {
      appendMessage("warning", `Failed to delete user - ${err}`);
    }
    closeDialog();
  }, [apiRequests, appendMessage, closeDialog, setLocation, user, username]);

  return (
    <div className="content mb-4" style="max-width: 65rem">
      <EditUserCard
        user={user}
        allUsersError={error}
        errors={errors}
        onDelete={requestDelete}
        onSave={saveChanges}
        setValue={setValue}
      />
      <DeleteUserDialog
        username={username}
        onCancel={closeDialog}
        onConfirm={onConfirmDelete}
      />
    </div>
  );
}
