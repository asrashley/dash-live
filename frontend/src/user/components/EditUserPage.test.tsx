import { afterEach, beforeEach, describe, expect, test, vi } from "vitest";
import fetchMock from "@fetch-mock/vitest";
import userEvent from "@testing-library/user-event";
import { mock } from "vitest-mock-extended";
import { useLocation, useParams } from "wouter-preact";
import log from "loglevel";

import { routeMap, uiRouteMap } from "@dashlive/routemap";
import { renderWithProviders } from "../../test/renderWithProviders";
import { ApiRequests, EndpointContext } from "../../endpoints";
import EditUserPage from "./EditUserPage";

import { FakeEndpoint, jsonResponse } from "../../test/FakeEndpoint";
import {
  adminUser,
  mediaUser,
  MockDashServer,
  normalUser,
  UserModel,
} from "../../test/MockServer";
import { useMessages, UseMessagesHook } from "../../hooks/useMessages";

vi.mock("wouter-preact", async (importOriginal) => {
  return {
    ...(await importOriginal()),
    useLocation: vi.fn(),
    useParams: vi.fn(),
  };
});

vi.mock("../../hooks/useMessages");

describe("EditUserPage component", () => {
  const needsRefreshToken = vi.fn();
  const hasUserInfo = vi.fn();
  const setLocation = vi.fn();
  const useParamsMock = vi.mocked(useParams);
  const useLocationMock = vi.mocked(useLocation);
  const useMessagesSpy = vi.mocked(useMessages);
  const useMessagesHook = mock<UseMessagesHook>();
  let endpoint: FakeEndpoint;
  let server: MockDashServer;
  let user: UserModel;
  let apiReq: ApiRequests;

  beforeEach(() => {
    log.setLevel("error");
    endpoint = new FakeEndpoint(document.location.origin);
    server = new MockDashServer({
      endpoint,
    });
    user = server.login(adminUser.username, adminUser.password);
    expect(user).not.toBeNull();
    apiReq = new ApiRequests({ needsRefreshToken, hasUserInfo });
    apiReq.setRefreshToken(user.refreshToken);
    apiReq.setAccessToken(user.accessToken);
    useParamsMock.mockReturnValue({
      username: mediaUser.username,
    });
    useLocationMock.mockReturnValue([
      uiRouteMap.editUser.url({ username: mediaUser.username }),
      setLocation,
    ]);
    useMessagesSpy.mockReturnValue(useMessagesHook);
  });

  afterEach(() => {
    vi.clearAllMocks();
    endpoint.shutdown();
    fetchMock.mockReset();
  });

  test("matches snapshot", async () => {
    const { asFragment, findByText } = renderWithProviders(
      <EndpointContext.Provider value={apiReq}>
        <EditUserPage />
      </EndpointContext.Provider>
    );
    await findByText(`Editing user ${mediaUser.username}`);
    expect(asFragment()).toMatchSnapshot();
  });

  test.each<boolean>([true, false])(
    "can edit a user, success=%s",
    async (success: boolean) => {
      const user = userEvent.setup();
      const { findByText, findBySelector, getByText } = renderWithProviders(
        <EndpointContext.Provider value={apiReq}>
          <EditUserPage />
        </EndpointContext.Provider>
      );
      const username = "new.username";
      const email = "new.username@test.local";
      const password = "s3cret!";
      const userInp = (await findBySelector(
        'input[name="username"]'
      )) as HTMLInputElement;
      await user.click(userInp);
      await user.clear(userInp);
      await user.type(userInp, success ? username : normalUser.username);
      expect(userInp.value).toEqual(success ? username : normalUser.username);
      const emailInp = (await findBySelector(
        'input[name="email"]'
      )) as HTMLInputElement;
      await user.click(emailInp);
      await user.clear(emailInp);
      await user.type(emailInp, email);
      const pwdInp = (await findBySelector(
        'input[name="password"]'
      )) as HTMLInputElement;
      await user.clear(pwdInp);
      await user.type(pwdInp, password);
      const confirmInp = (await findBySelector(
        'input[name="confirmPassword"]'
      )) as HTMLInputElement;
      await user.clear(confirmInp);
      await user.type(confirmInp, password);
      const editProm = endpoint.addResponsePromise(
        "post",
        routeMap.editUser.url({ upk: mediaUser.pk })
      );
      const btn = (await findByText("Save Changes")) as HTMLButtonElement;
      expect(btn.disabled).toEqual(!success);
      if (!success) {
        getByText("user already exists");
        await user.click(userInp);
        await user.clear(userInp);
        await user.type(userInp, username);
        expect(btn.disabled).toEqual(false);
        server.addUser({
          username,
          email,
        });
      }
      await user.click(btn);
      const result = await editProm;
      expect(result).toEqual(
        expect.objectContaining({
          status: 200,
          body: expect.any(String),
        })
      );
      const body = JSON.parse(result.body as string);
      if (success) {
        const newUser = server.getUser({ username });
        expect(newUser).toBeDefined();
        expect(newUser.password).toEqual(password);
        expect(body).toEqual({
          errors: [],
          success,
          user: {
            email,
            username,
            groups: ["USER", "MEDIA"],
            lastLogin: null,
            mustChange: false,
            pk: mediaUser.pk,
          },
        });
      } else {
        expect(body).toEqual({
          errors: [
            `${username} already exists`,
            `${email} email address already exists`,
          ],
          success,
        });
      }
    }
  );

  test("can delete a user", async () => {
    const user = userEvent.setup();
    const { findByText, findBySelector, getByTestId } = renderWithProviders(
      <EndpointContext.Provider value={apiReq}>
        <EditUserPage />
      </EndpointContext.Provider>
    );
    const delBtn = (await findByText("Delete User")) as HTMLButtonElement;
    await user.click(delBtn);
    await findBySelector(".modal-dialog");
    await findByText("Are you sure you want to delete this user?");
    const delProm = endpoint.addResponsePromise(
      "delete",
      routeMap.editUser.url({ upk: mediaUser.pk })
    );
    const btn = getByTestId("confirm-delete") as HTMLButtonElement;
    await user.click(btn);
    await expect(delProm).resolves.toEqual(
      expect.objectContaining({
        status: 204,
      })
    );
    expect(useMessagesHook.appendMessage).toHaveBeenCalledWith(
      "success",
      "User mediamgr successfully deleted"
    );
  });

  test("failing to delete a user", async () => {
    const user = userEvent.setup();
    const { findByText, findBySelector, getByTestId } = renderWithProviders(
      <EndpointContext.Provider value={apiReq}>
        <EditUserPage />
      </EndpointContext.Provider>
    );
    const delBtn = (await findByText("Delete User")) as HTMLButtonElement;
    await user.click(delBtn);
    await findBySelector(".modal-dialog");
    await findByText("Are you sure you want to delete this user?");
    const delProm = new Promise<void>((resolve) => {
      endpoint.setResponseModifier(
        "delete",
        routeMap.editUser.url({ upk: mediaUser.pk }),
        async () => {
          resolve();
          return jsonResponse("", 404);
        }
      );
    });
    const btn = getByTestId("confirm-delete") as HTMLButtonElement;
    await user.click(btn);
    await delProm;
    expect(useMessagesHook.appendMessage).toHaveBeenCalledWith(
      "warning",
      "Failed to delete user - Error: http://localhost:3000/api/users/101: 404"
    );
  });
});
