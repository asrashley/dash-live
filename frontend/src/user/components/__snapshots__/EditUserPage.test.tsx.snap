// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EditUserPage component > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="content mb-4"
    style="max-width: 65rem;"
  >
    <div
      class="card "
      id="edit-user"
    >
      <div
        class="card-header"
      >
        Editing user mediamgr
      </div>
      <div
        class="card-body"
      >
        <form>
          <div
            class="row mb-2 form-group "
          >
            <label
              class="col-2 col-form-label"
              for="model-username"
              id="label-username"
            >
              Username:
            </label>
            <div
              class="col-10"
            >
              <input
                aria-describedby="label-username"
                class="form-control is-valid "
                id="model-username"
                name="username"
                title="Username"
                type="text"
              />
            </div>
            <div
              class="invalid-feedback"
              style=""
            />
          </div>
          <div
            class="row mb-2 form-group "
          >
            <label
              class="col-2 col-form-label"
              for="model-email"
              id="label-email"
            >
              Email:
            </label>
            <div
              class="col-10"
            >
              <input
                aria-describedby="label-email"
                class="form-control is-valid "
                id="model-email"
                name="email"
                title="Email"
                type="email"
              />
            </div>
            <div
              class="invalid-feedback"
              style=""
            />
          </div>
          <div
            class="row mb-2 form-group "
          >
            <label
              class="col-2 col-form-label"
              for="model-password"
              id="label-password"
            >
              Password:
            </label>
            <div
              class="col-10"
            >
              <div
                class="input-group"
              >
                <input
                  aria-describedby="label-password"
                  class="form-control is-valid "
                  id="model-password"
                  name="password"
                  placeholder="***"
                  title="Password"
                  type="password"
                />
                <button
                  class="d-none"
                  type="button"
                >
                  Reveal
                </button>
              </div>
            </div>
            <div
              class="invalid-feedback"
              style=""
            />
          </div>
          <div
            class="row mb-2 form-group "
          >
            <label
              class="col-2 col-form-label"
              for="model-confirmPassword"
              id="label-confirmPassword"
            >
              Confirm Password:
            </label>
            <div
              class="col-10"
            >
              <div
                class="input-group"
              >
                <input
                  aria-describedby="label-confirmPassword"
                  class="form-control is-valid "
                  id="model-confirmPassword"
                  name="confirmPassword"
                  placeholder="***"
                  title="Confirm Password"
                  type="password"
                />
                <button
                  class="d-none"
                  type="button"
                >
                  Reveal
                </button>
              </div>
            </div>
            <div
              class="invalid-feedback"
              style="display: none;"
            />
          </div>
          <div
            class="row mb-2 form-group "
          >
            <label
              class="col-2 col-form-label"
              for="model-mustChange"
              id="label-mustChange"
            >
              Must Change:
            </label>
            <div
              class="col-7"
            >
              <input
                aria-describedby="text-mustChange"
                class="form-check-input is-valid "
                id="model-mustChange"
                name="mustChange"
                title="Must Change"
                type="checkbox"
              />
            </div>
            <div
              class="col-3 form-text"
              id="text-mustChange"
            >
              User must change password at next login?
            </div>
            <div
              class="invalid-feedback"
              style="display: none;"
            />
          </div>
          <div
            class="row mb-2 form-group "
          >
            <label
              class="col-2 col-form-label"
              for="model-groups"
              id="label-groups"
            >
              Groups:
            </label>
            <div
              class="col-10"
            >
              <div
                class="form-control is-valid "
                data-testid="msi-groups"
              >
                <div
                  class="form-check form-check-inline"
                >
                  <input
                    class="form-check-input"
                    id="msiadminGroup"
                    name="adminGroup"
                    type="checkbox"
                  />
                  <label
                    class="form-check-label me-3"
                    for="msiadminGroup"
                  >
                    ADMIN
                  </label>
                </div>
                <div
                  class="form-check form-check-inline"
                >
                  <input
                    class="form-check-input"
                    id="msimediaGroup"
                    name="mediaGroup"
                    type="checkbox"
                  />
                  <label
                    class="form-check-label me-3"
                    for="msimediaGroup"
                  >
                    MEDIA
                  </label>
                </div>
                <div
                  class="form-check form-check-inline"
                >
                  <input
                    class="form-check-input"
                    id="msiuserGroup"
                    name="userGroup"
                    type="checkbox"
                  />
                  <label
                    class="form-check-label me-3"
                    for="msiuserGroup"
                  >
                    USER
                  </label>
                </div>
              </div>
            </div>
            <div
              class="invalid-feedback"
              style="display: none;"
            />
          </div>
        </form>
        <div
          class="form-actions mt-2"
        >
          <button
            class="btn btn-primary me-3"
          >
            Save Changes
          </button>
          <button
            class="btn btn-danger me-3"
          >
            Delete User
          </button>
          <a
            class="btn btn-warning"
            href="/users"
          >
            Cancel
          </a>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
